pkg/sql/catalog/systemschema/system.go

```diff
@@ -248,6 +248,36 @@ CREATE TABLE system.jobs (
 	FAMILY claim (claim_session_id, claim_instance_id, num_runs, last_run)
 );`
 
+	// JobStatusTableSchema is the schema of the system.job_status table, which
+	// contains the "status" messages for each job, sorted by newest to oldest.
+	JobStatusTableSchema = `
+	CREATE TABLE system.job_status (
+		job_id INT8 NOT NULL,
+		kind STRING, -- the type of message, e.g. "state", "info", etc
+		written TIMESTAMPTZ NOT NULL DEFAULT now(), 
```

Comment:
I think you should consider removing the written field from the primary key of both of these tables. Including it forces the query optimizer to plan range scans instead of point lookups when joining against the table and we don't have a concrete need for the history. 

Reading through the comment on JobProgressTableSchema, I see this is motivated by a desire to avoid large amounts of mvcc history. I suspect the fact these rows are much smaller than the existing jobs table is enough to mitigate the history problem.

Reply:
a) the status table is actually keeping the history, and using `written` to do so.

b) progress _as it stands_ is not keeping it, but we have discussed wanting to do so, as it would be nice to see the "recent" history of the progress, i.e. is it stuck or moving.

c) Several years ago we declared that the jobs system had to move to one-row-per-update for any table that a job's own code could cause to be updated -- we could use one-row-per-job for the canonical set of jobs, but any table that a given job could cause writes to had to be mvcc-over-mvcc like this, due to the deficiency of our SQL execution engine w.r.t reading a row that has revisions split across ranges. We were bitten far too many times by ranges that became unavailable to _all writes_ because one job in the range had too many revisions of a record. Even with these rows being much smaller, a changefeed updating its timestamp over and over will eventually produce enough revisions to fill a range if GC is blocked, which is a thing that unfortunately happens for a variety of reasons. This aspect of the design -- mvcc-over-mvcc for these two tables -- was a pretty core assumption from the outset.

d)  "range scans instead of point lookups" can you say more here? We know the row will exist, so the ability of Get vs Scan wrt bloom filters doesn't apply, and after the filters a get is just a scan anyway since we don't know the mvcc ts.

Reply:
> a) the status table is actually keeping the history, and using written to do so.
> b) progress as it stands is not keeping it, but we have discussed wanting to do so, as it would be nice to see the "recent" history of the progress, i.e. is it stuck or moving.

I see we are keeping the history, but do we actually use this history? If code doesn't use the history, then logs or metrics are a better way to track it.

> c) Several years ago we declared that the jobs system had to move to one-row-per-update for any table that a job's own code could cause to be updated -- we could use one-row-per-job for the canonical set of jobs, but any table that a given job could cause writes to had to be mvcc-over-mvcc like this, due to the deficiency of our SQL execution engine w.r.t reading a row that has revisions split across ranges.

I would expect the SQL or KV layers to be responsible for mitigating large mvcc histories. If this is a stability problem for the job system, it could also impact user tables with high update rates. 

> d) "range scans instead of point lookups" can you say more here? We know the row will exist, so the ability of Get vs Scan wrt bloom filters doesn't apply, and after the filters a get is just a scan anyway since we don't know the mvcc ts.

In Cockroach Cloud, we have a table that has a many to one relationship where the many row is timestamped. Based on that experience, I think you will get the behavior you expect when you query the status for a single job: sql will plan a range scan with a limit, where the limit is pushed down to the scan operator. But if you try to join all jobs with the most recent status for each job, the query will plan a scan over the entire status history of each job, then a filter is applied after the join.

Reply:
> do we actually use this history

We have said it would be useful for years, so adding it was one of the goals here. I hope to see it on the job detail page at some point, but in the shortest term it would at least be in debug.zips.

> I would expect the SQL or KV layers to be responsible for mitigating large mvcc histories ... also impact user tables with high update rates.

Unfortunately SQL does not consider this a priority. It has indeed impacted user tables. The reason we're going out of our way in jobs is that jobs is shared infra and hosts many different jobs types in the same system, and some of those jobs are increasingly critical to cluster operations (eg span config recon) unlike a user table.

> the query will plan a scan over the entire status history of each job, then a filter is applied after the join.

We're already doing this join on the job_info table, since it has exactly this schema. The only difference here would be that _as it stands_, we'll retain more statuses per job than the info table currently retains progresses, however the info table can already retain many live keys per job due to dsp diag keys. I think the main thing is that there just aren't _that_ many keys per job that aren't relevant to the join. If the historical statuses per job do indeed become a problem, we can just add the same DELETE in the txn that info and progress use to only retain most recent; we don't need to change the schema or run a migration, though we might choose to pair that change with adding an extra status_history table to which we copy before deleting for the above-mentioned log on the detail page/debug.zips.